sync15-adapter/src/driver.rs

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// traits etc to drive a "full sync".
+
+use std::result;
+use failure;
+use client::{Sync15StorageClient, Sync15StorageClientInit};
+use state::{GlobalState, SetupStateMachine};
+use sync::{self, Store};
+use key_bundle::KeyBundle;
+
+
+pub trait GlobalStateProvider {
+    fn load(&self) -> result::Result<Option<GlobalState>, failure::Error>;
+
+    fn save(&self, state: Option<&GlobalState>) -> result::Result<(), failure::Error>;
+
+    // Store in memory, but do not persist, a Sync15StorageClientInit
+    fn get_client_init(&self) -> result::Result<Option<Sync15StorageClientInit>, failure::Error>;
+
+    fn set_client_init(&self, init: Option<Sync15StorageClientInit>) -> result::Result<(), failure::Error>;
+}
+
+pub fn sync(
+    store: &Store,
+    gsp: &GlobalStateProvider,
+    storage_init: &Sync15StorageClientInit,
+    root_sync_key: &KeyBundle
+) -> result::Result<(), failure::Error> {
+    let maybe_global = gsp.load()?;
+    // Note: We explicitly write a None back as the state, meaning if we
+    // unexpectedly fail below, the next sync will redownload meta/global,
+    // crypto/keys, etc. without needing to. Apparently this is both okay
+    // and by design.
+    gsp.save(None)?;
+    // `maybe_sync_info` is None if we haven't called `sync` since
+    // restarting the browser.
+    //
+    // If this is the case we may or may not have a persisted version of
+    // GlobalState stored in the DB (we will iff we've synced before, unless
+    // we've `reset()`, which clears it out).
+    // XXX - fix comment above?
+
+    let mut global_state = match maybe_global {
+        Some(g) => g,
+        None => {
+            info!("First time through since unlock. Creating default global state.");
+            gsp.set_client_init(None)?;
+            GlobalState::default()
+        }
+    };
+
+    // XXX - storage_init.
+//            };
+        let client = Sync15StorageClient::new(storage_init.clone())?;
```

Comment:
We really don't want to be recreating this every time. Was there a reason you removed the other logic?